pkg/ebpf/network_filter.go

```diff
@@ -0,0 +1,208 @@
+package ebpf
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// ConnectionFilter holds a user-defined blacklisted IP/CIDR, and ports
+type ConnectionFilter struct {
+	IP       *net.IPNet
+	Ports    map[uint16]ConnTypeFilter
+	AllPorts ConnTypeFilter
+}
+
+// ConnTypeFilter holds user-defined protocols
+type ConnTypeFilter struct {
+	TCP bool
+	UDP bool
+}
+
+// ParseConnectionFilters takes the user defined blacklist and returns a slice of ConnectionFilters
+func ParseConnectionFilters(filters map[string][]string) (blacklist []*ConnectionFilter) {
+	for ip, ports := range filters {
+		filter := &ConnectionFilter{Ports: map[uint16]ConnTypeFilter{}}
+		var subnet *net.IPNet
+		var err error
+
+		// retrieve valid IPs
+		if strings.ContainsRune(ip, '*') {
+			subnet = nil // use for wildcard
+		} else if strings.ContainsRune(ip, '/') {
+			_, subnet, err = net.ParseCIDR(ip)
+		} else if strings.ContainsRune(ip, '.') {
+			_, subnet, err = net.ParseCIDR(ip + "/32") // if given ipv4, prefix length of 32
+		} else if strings.Contains(ip, "::") {
+			_, subnet, err = net.ParseCIDR(ip + "/64") // if given ipv6, prefix length of 64
+		} else {
+			log.Errorf("Invalid IP/CIDR/* defined for connection filter: %s", err)
+			continue
+		}
+
+		if err != nil {
+			log.Errorf("Given filter will not be respected. Could not parse given IPs: %s", err)
+			continue
+		}
+
+		filter.IP = subnet
+		var validFilter bool
+		for _, v := range ports {
+			lowerPort, upperPort, connTypeFilter, valid := parsePortAndProtocolFilter(v)
+			validFilter = valid
+
+			if !validFilter {
+				break
+			}
+
+			// port is wildcard
+			if lowerPort == 0 && upperPort == 0 {
+				filter.AllPorts.TCP = connTypeFilter.TCP || filter.AllPorts.TCP
```

Comment:
Can you share an example where `filter.AllPorts.TCP` was already true here? This is pretty odd to me

Reply:
Since IP/CIDR * and port * result in an error, was trying to prevent a configuration like this from being considered valid:
```
system_probe_config.source_excludes: 
  "*":
    - "TCP *"
    - "UDP *"
```